lexos/models/kmeans_model.py

```diff
@@ -0,0 +1,455 @@
+        result_table["Document"] = labels
+        result_table["X-Coordinate"] = reduced_data[:, 0]
+        result_table["Y-Coordinate"] = reduced_data[:, 1]
+
+        return result_table
+
+    def _get_3d_frame(self, k_means_index: List[int]) -> pd.DataFrame:
+        """Generate 3 dimensional table result for K-Means analysis.
+
+        :param k_means_index: the cluster result of K-Means analysis.
+        :return: A pandas data frame with four columns.The first column
+        contains cluster numbers and the second column contains document names,
+        the rest columns contain the coordinates of the files.
+        """
+        # Get reduced data.
+        reduced_data = self._get_reduced_data()
+        # Get file names.
+        labels = [self._id_temp_label_map[file_id]
+                  for file_id in self._doc_term_matrix.index.values]
+
+        # Initialize the table with proper headers.
+        result_table = pd.DataFrame(columns=["Cluster #",
+                                             "Document",
+                                             "X-Coordinate",
+                                             "Y-Coordinate",
+                                             "Z-Coordinate"])
+
+        # Fill the pandas data frame.
+        result_table["Cluster #"] = [index + 1 for index in k_means_index]
+        result_table["Document"] = labels
+        result_table["X-Coordinate"] = reduced_data[:, 0]
+        result_table["Y-Coordinate"] = reduced_data[:, 1]
+        result_table["Z-Coordinate"] = reduced_data[:, 2]
+
+        return result_table
+
+    def _get_voronoi_result(self) -> KMeansUnprocessedResult:
+        """Generate voronoi formatted graph for K Means result.
+
+        :return: A plotly object hat has been converted to HTML format string.
+        """
+        # Get kMeans analyze result and unpack it.
+        k_means = self._get_k_means()
+        reduced_data = self._get_reduced_data()
+        k_means_index = k_means.fit_predict(reduced_data)
+
+        # This is important, such that plot and table results are consistent.
+        sorted_k_means_unique_index = sorted(set(k_means_index))
+
+        # Get file names.
+        labels = np.array([self._id_temp_label_map[file_id]
+                           for file_id in self._doc_term_matrix.index.values])
+
+        # Get a list of lists of file names based on the cluster result.
+        cluster_labels = [labels[np.where(k_means_index == index)]
+                          for index in sorted_k_means_unique_index]
+
+        # Get a list of lists of file coordinates based on the cluster result.
+        cluster_values = [reduced_data[np.where(k_means_index == index)]
+                          for index in sorted_k_means_unique_index]
+
+        # Get a list of centroid results based on the cluster result.
+        centroid_values = [np.mean(cluster, axis=0, dtype="float_")
+                           for cluster in cluster_values]
```

Comment:
I think these are three parallel arrays, remember we conceal parallel arrays because they are bug prone. You can turn them into one panda data frame or two panda series.